fe2-android/app/src/androidTest/java/com/github/dedis/popstellar/ui/detail/event/consensus/ElectionStartFragmentTest.java

```diff
@@ -0,0 +1,278 @@
+
+  private static final Lao lao = new Lao(laoId);
+  private static final Election election = new Election(laoId, pastTime, electionName);
+  private static final String node2Key = "12RDUW4s9bZrdqJB7zoVag";
+  private static final String node3Key = "dRCXFdXYfY5OVAzh3oQ3pA";
+  private static final ConsensusKey key = new ConsensusKey("election", election.getId(), "state");
+  private static final String instanceId =
+      Consensus.generateConsensusId(key.getType(), key.getId(), key.getProperty());
+
+  private static final ConsensusElect elect =
+      new ConsensusElect(pastTime, key.getId(), key.getType(), key.getProperty(), "started");
+  private static final ConsensusElectAccept accept3 =
+      new ConsensusElectAccept(instanceId, "m3", true);
+
+  private static final ConsensusLearn learn3 =
+      new ConsensusLearn(instanceId, "m3", Collections.emptyList());
+
+  private static final ArgumentCaptor<Message> captor = ArgumentCaptor.forClass(Message.class);
+
+  private static String publicKey;
+  private static LAORepository laoRepository;
+  private static boolean isAlreadySetup = false;
+
+  LAORemoteDataSource remoteDataSource = Mockito.mock(LAORemoteDataSource.class);
+  LAOLocalDataSource localDataSource = Mockito.mock(LAOLocalDataSource.class);
+
+  @Rule
+  public final FragmentScenarioRule<ElectionStartFragment> fragmentRule =
+      FragmentScenarioRule.launchInContainer(
+          ElectionStartFragment.class,
+          () -> {
+            ElectionStartFragment fragment = ElectionStartFragment.newInstance();
+            if (!isAlreadySetup) {
+              setup();
+              isAlreadySetup = true;
+            }
+            return fragment;
+          });
+
+  private void setup() {
+    Mockito.when(remoteDataSource.incrementAndGetRequestId()).thenReturn(42);
+    Mockito.when(remoteDataSource.observeWebsocket()).thenReturn(Observable.empty());
+    Observable<GenericMessage> upstream = Observable.fromArray((GenericMessage) new Result(42));
+
+    // Mock the remote data source to receive a response
+    Mockito.when(remoteDataSource.observeMessage()).thenReturn(upstream);
+
+    try {
+      LAORepository.destroyInstance();
+      laoRepository =
+          LAORepository.getInstance(
+              remoteDataSource,
+              localDataSource,
+              Injection.provideAndroidKeysetManager(ApplicationProvider.getApplicationContext()),
+              Injection.provideGson(),
+              new ProdSchedulerProvider());
+      publicKey = getPublicKey();
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+
+    election.setStart(futureTime);
+    lao.setChannel(laoChannel);
+    lao.setOrganizer(publicKey);
+    lao.setWitnesses(Sets.newSet(node2Key, node3Key));
+
+    laoRepository.getLaoById().put(laoChannel, new LAOState(lao));
+    laoRepository.updateNodes(laoChannel);
+  }
+
+  @After
+  public void clean() {
+    LAORepository.destroyInstance();
+  }
```

Comment:
The @After is called after every tests.

As there is only one, this is not a big issue. But we might want to pass to JUnit 5 to get access to @BeforeAll and @AfterAll.

Or, you can create and destroy the LAORepository instance for each tests with @Before and @After, which makes them completely independent.

Reply:
I made a single test because most of the test is sequential and need the previous updates to work.
The setup function need to be called before the fragment is fully initialized, so it can't be used with @before.
And the clean function is used only to be sure that it won't interfere with test in others files.